src/hotspot/share/opto/library_call.cpp

```diff
@@ -6967,6 +6969,77 @@ bool LibraryCallKit::inline_base64_decodeBlock() {
   return true;
 }
 
+bool LibraryCallKit::inline_poly1305_processBlocks() {
+  address stubAddr;
+  const char *stubName;
+  assert(UsePolyIntrinsics, "need Poly intrinsics support");
+  assert(callee()->signature()->size() == 3, "poly1305_processBlocks has %d parameters", callee()->signature()->size());
+  stubAddr = StubRoutines::poly1305_processBlocks();
+  stubName = "poly1305_processBlocks";
+
+  if (!stubAddr) return false;
+  Node* polyObj = argument(0);
+  Node* input = argument(1);
+  Node* input_offset = argument(2);
+  Node* len = argument(3);
+
+  Node* accFace = load_field_from_object(polyObj, "a", "Lsun/security/util/math/MutableIntegerModuloP;");
+  assert(accFace != NULL, "Accumulator field is null");
+  const TypeInstPtr* ainst = _gvn.type(accFace)->isa_instptr();
+  assert(ainst != NULL, "Accumulator obj is null");
+  assert(ainst->is_loaded(), "MutableIntegerModuloP obj is not loaded");
+  ciKlass* klass_MutableElement = ainst->instance_klass()->find_klass(ciSymbol::make("sun/security/util/math/intpoly/IntegerPolynomial$MutableElement"));
+  assert(klass_MutableElement != NULL, "IntegerPolynomial$MutableElement class is null");
+  assert(klass_MutableElement->is_loaded(), "IntegerPolynomial$MutableElement class is not loaded");
+  ciInstanceKlass* instklass_MutableElement = klass_MutableElement->as_instance_klass();
+
+  const TypeKlassPtr* aklass = TypeKlassPtr::make(instklass_MutableElement);
+  const TypeOopPtr* atype = aklass->as_instance_type()->cast_to_ptr_type(TypePtr::NotNull);
+  Node* accObj = new CheckCastPPNode(control(), accFace, atype);
+  accObj = _gvn.transform(accObj);
+  Node* alimbs = load_field_from_object(accObj, "limbs", "[J");
+
+  Node* rFace = load_field_from_object(polyObj, "r", "Lsun/security/util/math/IntegerModuloP;"); //this.r.limbs
+  assert(rFace != NULL, "R field is null");
+  const TypeInstPtr* rinst = _gvn.type(rFace)->isa_instptr();
+  assert(rinst != NULL, "R obj is null");
+  assert(rinst->is_loaded(), "IntegerModuloP obj is not loaded");
+  ciKlass* klass_ImmutableElement = rinst->instance_klass()->find_klass(ciSymbol::make("sun/security/util/math/intpoly/IntegerPolynomial$ImmutableElement"));
+  assert(klass_ImmutableElement != NULL, "IntegerPolynomial$ImmutableElement class is null");
+  assert(klass_ImmutableElement->is_loaded(), "IntegerPolynomial$ImmutableElement class is not loaded");
+  ciInstanceKlass* instklass_ImmutableElement = klass_ImmutableElement->as_instance_klass();
+
+  const TypeKlassPtr* rklass = TypeKlassPtr::make(instklass_ImmutableElement);
+  const TypeOopPtr* rtype = rklass->as_instance_type()->cast_to_ptr_type(TypePtr::NotNull);
+  Node* rObj = new CheckCastPPNode(control(), rFace, rtype);
```

Comment:
FTR it's an unsafe cast since it  doesn't involve a runtime check from `IntegerModuloP` to `ImmutableElement`. Please, lift as much checks into Java wrapper as possible.

Reply:
Ah, yeah.. I quite suspected I didn't emulate all the bytecodes needed. Thanks for the info.

So this is a bit of a quandary.. I had done the intrinsic more in Java before, but it slows down the non-intrinsic path (This was the discussion we were having with Jamil).

In Java, the limbs are not 'accessible' per-se.. They are in a separate package hidden behind and interface.. and in a nested non-static class inside an abstract class... Its quite well designed. Its just makes what I want to do break most encapsulations.

There is a method (`asByteArray`) to extract the limbs that I was previously using, but that slows down non-intrinsic path (to be honest it slows down the intrinsic path too, but the assembler makes some of that back.. You can see that from the numbers I posted for Jamil, vs original in the PR header; originally I got 18x, and now with accessing limbs directly its 19x). 

If I had some way to check 'is intrinsic available', I could at least not slow down the current code. I would still have to break the encapsulation to do the checks/casts though. It all seems less-then-perfect.

Reply:
@iwanowww just to save some of your time...  Sandhya suggested another way to move the checks to Java, hopefully without too much penalty to non-intrinsic path. Should upload it later today.

Reply:
@iwanowww Please have a look, just pushed a different way to fetch the limbs.